### PR TITLE
Drop support for LLVM <= 3.9.1

### PR DESCRIPTION
**Pull request by Keno**
As of the merge of #21888, we no longer support building on LLVM <= 3.9.1.
This is essentialyl the mechanical cleanup to rip out all code that
we had to support building on older LLVM versions. There may still be
some residual support left in places and of course some things can now
be cleaned up further, but this should get us started.

### CODE REVIEW

File: src/ccall.cpp [hunk 1]
```diff
@@ -1136,20 +1101,14 @@ static jl_cgval_t emit_llvmcall(jl_value_t **args, size_t nargs, jl_codectx_t *c
             it != argtypes.end(); ++it, ++i)
             assert(*it == f->getFunctionType()->getParamType(i));
 
-#ifdef USE_MCJIT
```
Comment (ararslan) [inline src/ccall.cpp]: Are we assuming MCJIT now rather than ORC?

File: src/llvm-late-gc-lowering.cpp [hunk 2]
```diff
@@ -583,8 +583,7 @@ static int NoteSafepoint(State &S, BBState &BBS, CallInst *CI) {
 }
 
 void LateLowerGCFrame::NoteUse(State &S, BBState &BBS, Value *V, BitVector &Uses) {
-    // Short circuit to have to avoid dealing specially with vectors of
-    // constants, etc.
+    // Short circuit to avoid having to dealing with vectors of constants, etc.
```
Comment (tkelman) [inline src/llvm-late-gc-lowering.cpp]: having to deal with

**Comment by tkelman** [comment]
not yet addressed:
https://github.com/JuliaLang/julia/pull/21888#discussion_r122405141
https://github.com/JuliaLang/julia/pull/21888#discussion_r120031311

**Comment by Keno** [comment]
I typo'd here. This drops support for < 3.9.1 or <= 3.9.0

File: src/ccall.cpp (hunk 1 above)
Comment (yuyichao) [inline src/ccall.cpp]: USE_MCJIT is always defined for `USE_ORCJIT`.

**Comment by tkelman** [comment]
also delete old patches and build system code for these old versions?

File: src/codegen.cpp [hunk 3]
```diff
@@ -4842,16 +4611,12 @@ static Function *gen_cfun_wrapper(jl_function_t *ff, jl_value_t *jlrettype, jl_t
                 }
                 (void)at;
             }
-            Value *argn = builder.CreateConstInBoundsGEP1_32(LLVM37_param(NULL) myargs, i);
+            Value *argn = builder.CreateConstInBoundsGEP1_32(NULL, myargs, i);
```
Comment (vtjnash) [inline src/codegen.cpp]: T_prjlvalue

File: src/codegen.cpp [hunk 4]
```diff
@@ -4710,7 +4487,7 @@ static Function *gen_cfun_wrapper(jl_function_t *ff, jl_value_t *jlrettype, jl_t
         // for jlcall, we need to pass the function object even if it is a ghost.
         // here we reconstruct the function instance from its type (first elt of argt)
         Value *theF = literal_pointer_val((jl_value_t*)ff);
-        GetElementPtrInst *slot = GetElementPtrInst::Create(LLVM37_param(NULL) myargs,
+        GetElementPtrInst *slot = GetElementPtrInst::Create(NULL, myargs,
```
Comment (vtjnash) [inline src/codegen.cpp]: T_prjlvalue

File: src/codegen.cpp [hunk 5]
```diff
@@ -4725,22 +4502,14 @@ static Function *gen_cfun_wrapper(jl_function_t *ff, jl_value_t *jlrettype, jl_t
             builder.CreateCondBr(age_ok, b_jlcall, b_generic);
             builder.SetInsertPoint(b_jlcall);
             Value *nargs_v = ConstantInt::get(T_int32, nargs);
-            Value *myargs1 = builder.CreateConstInBoundsGEP1_32(LLVM37_param(NULL) myargs, 1);
-#if JL_LLVM_VERSION >= 30700
+            Value *myargs1 = builder.CreateConstInBoundsGEP1_32(NULL, myargs, 1);
```
Comment (vtjnash) [inline src/codegen.cpp]: T_prjlvalue

File: src/codegen.cpp [hunk 6]
```diff
@@ -4628,7 +4407,7 @@ static Function *gen_cfun_wrapper(jl_function_t *ff, jl_value_t *jlrettype, jl_t
         // figure out how to repack this type
         if (!specsig) {
             Value *arg = boxed(inputarg, &ctx, false); // don't want a gcroot, since it's about to be put into the jlcall frame anyways
-            GetElementPtrInst *slot = GetElementPtrInst::Create(LLVM37_param(NULL) myargs,
+            GetElementPtrInst *slot = GetElementPtrInst::Create(NULL, myargs,
```
Comment (vtjnash) [inline src/codegen.cpp]: T_prjlvalue

File: src/ccall.cpp (hunk 1 above)
Comment (Keno) [inline src/ccall.cpp]: Yes, the default use to be the very old (3.3) JIT. Now MCJIT is the default.

**Comment by Keno** [comment]
Since this touches so much code, I'll go ahead and merge this and do follow-up cleanups in a separate PR to avoid having to rebase.

File: src/ccall.cpp (hunk 1 above)
Comment (ararslan) [inline src/ccall.cpp]: Ah, thanks for the explanation. Didn't know.

**Pull request merged**